bloom.go

```diff
@@ -0,0 +1,310 @@
+var bloomHasherIds = map[uintptr]byte{
+	reflect.ValueOf(bloomHasherCRCBlocked64B8K3).Pointer(): 1,
+}
+
+// bloomHashers maps from hash identifierss stored in encoded bloom filters to
+// hash functions, to allo backwards compatible hash function evolution.
+var bloomHashers = []bloomHash{
+	bloomHasherCRCBlocked64B8K3,
+}
+
+// The following functions and constants *must not* be changed unless you can prove
+// they have exactly identical behavior. Instead of changing these functions,
+// add a new hash function and a new entry in bloomHasherIds and bloomHashers,
+// then change the default hash function.
+//
+// This allows changing to a new hash function without invalidating all existing
+// files, and more importantly, without starting to return false negatives (!!!)
+// because the hash function changed unexpectedly.
+const bloomHashMinWordLength = 4
+
+// bloomWordTab uses a table to implement a matcher for the regex \w{4,}
+var bloomWordTab [128 / 64]uint64 = initBloomWordTab()
+
+func initBloomWordTab() [128 / 64]uint64 {
+	var tab [128 / 64]uint64
+	for x := byte(0); x < 128; x++ {
+		if x == '_' || 'a' <= x && x <= 'z' || 'A' <= x && x <= 'Z' || '0' <= x && x <= '9' {
+			tab[x/64] |= 1 << (x % 64)
+		}
+	}
+	return tab
+}
+
+func findNextWord(i int, in []byte) (int, []byte) {
+	// Dropping the unicode case-folding requirement would
+	// improve performance here. There are *exactly* two Unicode
+	// codepoints that map down to ASCII:
+	//   K: U+212A KELVIN SIGN
+	//   ſ: U+017F LATIN SMALL LETTER LONG S
+	for i < len(in) {
+		// skip non-word runes
+		for i < len(in) {
+			c, sz := utf8.DecodeRune(in[i:])
+			c = unicode.ToLower(c)
+			if c < 128 && bloomWordTab[c/64]&(1<<(c%64)) != 0 {
+				break
+			}
+			i += sz
+		}
+		// count length of word section
+		wordStart := i
+		runeLength := 0
+		for i < len(in) {
+			c, sz := utf8.DecodeRune(in[i:])
+			c = unicode.ToLower(c)
+			if c >= 128 || bloomWordTab[c/64]&(1<<(c%64)) == 0 {
+				break
+			}
+			runeLength++
+			i += sz
+		}
+		// Skip short words.
+		if runeLength < bloomHashMinWordLength {
+			continue
+		}
+		return i, bytes.ToLower(in[wordStart:i])
+	}
+	return i, nil
+}
+
+func bloomHasherCRC(in []byte) []uint32 {
+	out := []uint32{}
+	// for i, s := findNextWord(0, in); s != nil; i, s = findNextWord(i, s) {
+	for i := 0; i < len(in); {
+		var s []byte
+		i, s = findNextWord(i, in)
+		// Add all substrings of length 4-10 to the bloom filter.
+		// Not having a bound on the maximum length causes quadratic
+		// probe counts on long "words"-- like a 241KB line of
+		// DNA ("gtggcaccctgactgg...")
+		for l := 10; l >= 4; l-- {
+			if l > len(s) {
+				continue
+			}
+			for i := 0; i+l <= len(s); i++ {
+				if '0' <= s[i] && s[i] <= '9' {
+					// Long numeric/hex constants are generally unlikely
+					// to be searched for, so don't include probes for
+					// substrings that start with a number.
+					continue
+				}
+				out = append(out, crc32.Checksum(s[i:i+l], crcTab))
```

Comment:
any benefit in using a stateful hasher so we can reuse the hash computed for `s[i:i+l-1]`? Or is it super minor given 4 <= l <= 10 and how fast crc is.

Reply:
Super minor, since the crc here is using hardware acceleration that takes <1 cycle/byte.